common/lib/container-definitions/api-report/container-definitions.api.md

```diff
@@ -129,7 +129,7 @@ export interface IContainer extends IEventProvider<IContainerEvents>, IFluidRout
     // @alpha
     readonly clientId?: string | undefined;
     close(error?: ICriticalContainerError): void;
-    closeAndGetPendingLocalState(): string;
+    closeAndGetPendingLocalState(): Promise<string>;
```

Comment:
I understand why you need to do so, but worth pointing out that it opens possibility for bugs (depends on design).
While this operation is in flight, user may continue to interact with application and make local edits. There could be 3 outcomes:
1. We capture state at the beginning of this operation. All such edits are ignored
   - This might be Ok if we app is Ok with ensuring user edits are disabled while this process is in flight. Even so, not great design.
2. We capture some of these edits (or rather parts of changes to data model), possibly leaving app data mode in inconsistent state
3. We capture all such edits, the state returned represents state of data model at the end of the process.

I'd love to understand better where we are landing here

Reply:
Good point. The async work in runtime is not really state dependent. Would it be sufficient to await reading the snapshot blobs then synchronously return state from runtime, then container?

I'm not sure if there's a meaningful difference between 1 and 3 though, since the blobs are cached and we are not actually awaiting network calls, the window for users to make edits should be pretty small, right? It could be different outside of ODSP if the blobs are not cached.

Reply:
The issue in JS is that even if async function does its work synchronously, it returns promise and caller does not resume work right away - instead continuation is scheduled, and work will continue at some point in future, allowing interruption of flow (only micro-tasks can creep in, but it's still a problem).
So, I think it will be very hard to have a design where we capture most of the container state synchronously, and thus avoid issues where state captured represents different points in time (and thus has different set of local changes).

I wonder if we should reconsider our approach here. I.e., force app to capture most of the state upfront (blobs), and then have synchronous flow to capture deltas? App controls when first part is called, and changes are allowed in between these two parts, and actually second part can be called multiple times (as time moves forward), but container state is combination of those two parts.

Reply:
Maybe it would be better to read the blobs in runtime load?

It would not really be safe to call the second part without closing, so it couldn't be called multiple times

Reply:
Possible. I'd point out that it's the expense that app have to pay whether they use serialization later or not, so it has some negatives.